PYME/Acquire/Hardware/pco/pco_cam.py

```diff
@@ -68,25 +74,21 @@ def CamReady(self):
         return self.initalized
 
     def ExtractColor(self, chSlice, mode):
-
         # Somehow this check matters... shouldn't this be taken care of by ExpReady???
-        if self.GetNumImsBuffered() < 1:
+        if (not self.recording) or (self.GetNumImsBuffered() < 1):
             return True
 
         if self.n_read < self.buffer_size:
             curr_frame = self.n_read
         else:
             curr_frame = self.buffer_size-self.GetNumImsBuffered()
 
-        # print('n_buffered: {}, n_read: {}, curr_frame: {}'.format(self.GetNumImsBuffered(), self.n_read, curr_frame))
-
         # Grab the image (unused metadata _)
         image, _ = self.cam.image(curr_frame)
 
-        # print('curr_frame: {}, recorder_image_number: {}, num_ims_buffered: {}, num_ims_read: {}'.format(curr_frame, meta['recorder image number'], self.cam.rec.get_status()['dwProcImgCount'], self.n_read))
-
         ctypes.cdll.msvcrt.memcpy(chSlice.ctypes.data_as(ctypes.POINTER(ctypes.c_uint16)),
-                   image.ctypes.data_as(ctypes.POINTER(ctypes.c_uint16)), chSlice.nbytes)
+                   image.ctypes.data_as(ctypes.POINTER(ctypes.c_uint16)), 
+                   self.__image_frame_bytes)
```

Comment:
This is super unsafe and likely to generate segfaults and/or corruption as it introduces the possibility of copying / writing over the end of the chSlice array. Because you are copying **into** `chSlice`, it **must** be the arbiter of the amount to be copied.   

You could theoretically get read segfaults on the camera image buffer with the old code if the sizes didn't match, but these are safer than writing to memory you don't own. For safety, consider the following:

- revert to old code
- add a line above such as:

```python
assert(self.__image_frame_bytes == chSlice.nbytes)
```

which will throw an exception if things are out of sync, but should prevent segfaults / corruption.

Reply:
Interesting. I originally had `self.__image_frame_bytes` as `chSlice.nbytes`, but I changed it because the `self.__image_frame_bytes` approach is what we do in `HamamatsuORCA`, so I assumed this was the preferred system.

Either way, a memory error here (in `memcpy`) is caught. I know, I've made lots of them.

Reply:
I did push a commit making this change, though.

Reply:
I  guess we should change the Orca code to0 - in general the Zyla is a (much) safer reference than the Orca.

If `memcpy` is catching things, it can only catch out of process memory issues - you still risk over-writing large chunks of in-process memory and causing data (and potentially even bytecode) corruption (it's completely unaware of python-internal memory allocation).

Reply:
Probably true, but every time I've made a memory error here only ExtractColor gets messed up. There's no segfaults, just a bunch of tracebacks telling me I screwed up and the program keeps rolling until I quit. It's a very noticeable, repeatable error, and it tends to happen right away rather than 10 minutes into imaging (which is the error I experience with the z-stacking with a low number of frames to average with this camera).

Reply:
The `Zyla` uses `chSlice` size, but not `nbytes`. Should be fairly similar to `__image_frame_bytes` actually, since `chSlice` size gets set by the chip/roi size.